govcd/metadata_test.go

```diff
@@ -0,0 +1,181 @@
+// +build vapp vdc metadata functional ALL
+
+/*
+ * Copyright 2018 VMware, Inc.  All rights reserved.  Licensed under the Apache v2 License.
+ */
+
+package govcd
+
+import (
+	"fmt"
+
+	. "gopkg.in/check.v1"
+)
+
+func init() {
+	testingTags["metadata"] = "metadata_test.go"
+}
+
+func (vcd *TestVCD) Test_AddMetadataForVdc(check *C) {
+	if vcd.config.VCD.Vdc == "" {
+		check.Skip("skipping test because no vDC name is empty")
+	}
+
+	fmt.Printf("Running: %s\n", check.TestName())
+
+	// Add metadata
+	task, err := vcd.vdc.AddMetadata("key", "value")
+	check.Assert(err, IsNil)
+	err = task.WaitTaskCompletion()
+	check.Assert(err, IsNil)
+	check.Assert(task.Task.Status, Equals, "success")
+
+	// Check if metadata was added correctly
+	metadata, err := vcd.vdc.GetMetadata()
+	check.Assert(err, IsNil)
+	check.Assert(metadata.MetadataEntry[0].Key, Equals, "key")
+	check.Assert(metadata.MetadataEntry[0].TypedValue.Value, Equals, "value")
+}
+
+func (vcd *TestVCD) Test_DeleteMetadataForVdc(check *C) {
+	if vcd.config.VCD.Vdc == "" {
+		check.Skip("skipping test because no vDC name is empty")
+	}
+
+	fmt.Printf("Running: %s\n", check.TestName())
+
+	// Add metadata
+	task, err := vcd.vdc.AddMetadata("key2", "value2")
+	check.Assert(err, IsNil)
+	err = task.WaitTaskCompletion()
+	check.Assert(err, IsNil)
+	check.Assert(task.Task.Status, Equals, "success")
+
+	// Remove metadata
+	task, err = vcd.vdc.DeleteMetadata("key2")
+	check.Assert(err, IsNil)
+	err = task.WaitTaskCompletion()
+	check.Assert(err, IsNil)
+	check.Assert(task.Task.Status, Equals, "success")
+	metadata, err := vcd.vdc.GetMetadata()
+	check.Assert(err, IsNil)
+	for _, k := range metadata.MetadataEntry {
+		if k.Key == "key2" {
+			check.Errorf("metadata.MetadataEntry should not contain key: %s", k)
+		}
+	}
+}
+
+func (vcd *TestVCD) Test_AddMetadataOnVapp(check *C) {
+	if vcd.skipVappTests {
+		check.Skip("Skipping test because vapp was not successfully created at setup")
+	}
+	// Add metadata
+	task, err := vcd.vapp.AddMetadata("key", "value")
+	check.Assert(err, IsNil)
+	err = task.WaitTaskCompletion()
+	check.Assert(err, IsNil)
+	check.Assert(task.Task.Status, Equals, "success")
+
+	// Check if metadata was added correctly
+	metadata, err := vcd.vapp.GetMetadata()
+	check.Assert(err, IsNil)
+	check.Assert(metadata.MetadataEntry[0].Key, Equals, "key")
```

Comment:
Shall we check whether `metadata.MetadataEntry` is a valid array before addressing an index?

Reply:
fixed